diff --git a/src/SH/sh.c b/src/SH/sh.c
index 46a1b80..731c536 100644
--- a/src/SH/sh.c
+++ b/src/SH/sh.c
@@ -222,7 +222,13 @@ void sh(struct waveSH *waveSH, struct waveSH_PML *waveSH_PML, struct matSH *matS
 				    }  
 
 				    if(muss>0.0){			
+#if defined(M5_SH_GRAD_X_ONLY)
+				         (*fwiSH).waveconv_u_shot[j][i] += (*fwiSH).forward_prop_sxz[imat] * (*waveSH).psxz[j][i];
+#elif defined(M5_SH_GRAD_Y_ONLY)
+				         (*fwiSH).waveconv_u_shot[j][i] += (*fwiSH).forward_prop_syz[imat] * (*waveSH).psyz[j][i];
+#else
 				         (*fwiSH).waveconv_u_shot[j][i] += ((*fwiSH).forward_prop_syz[imat] * (*waveSH).psyz[j][i]) + ((*fwiSH).forward_prop_sxz[imat] * (*waveSH).psxz[j][i]);
+#endif
 				    } 		                  
 		                }
 				
@@ -238,7 +244,13 @@ void sh(struct waveSH *waveSH, struct waveSH_PML *waveSH_PML, struct matSH *matS
 				    }  
 
 				    if(muss>0.0){			
+#if defined(M5_SH_GRAD_X_ONLY)
+				         (*fwiSH).waveconv_u_shot[j][i] += (*fwiSH).forward_prop_sxz[imat] * (*waveSH).psxz[j][i];
+#elif defined(M5_SH_GRAD_Y_ONLY)
+				         (*fwiSH).waveconv_u_shot[j][i] += (*fwiSH).forward_prop_syz[imat] * (*waveSH).psyz[j][i];
+#else
 				         (*fwiSH).waveconv_u_shot[j][i] += ((*fwiSH).forward_prop_syz[imat] * (*waveSH).psyz[j][i]) + ((*fwiSH).forward_prop_sxz[imat] * (*waveSH).psxz[j][i]);
+#endif
 				    } 		                  
 		                }			
 						                                                                                                     
